src/sock.c

```diff
@@ -485,18 +489,31 @@ ssize_t RETRACE_IMPLEMENTATION(send)(int sockfd, const void *buf, size_t len, in
 		errno = err;
 		ret = -1;
 	} else {
-		if (rtr_str_inject(STRINJECT_FUNC_SEND, (void *) buf, len)) {
+		if (rtr_str_inject(STRINJECT_FUNC_SEND, buf, len, &inject_buffer, &inject_len)) {
 			event_info.extra_info = "[redirected]";
 			event_info.event_flags = EVENT_FLAGS_PRINT_RAND_SEED | EVENT_FLAGS_PRINT_BACKTRACE;
 			event_info.logging_level |= RTR_LOG_LEVEL_FUZZ;
-		}
 
-		rtr_http_sniff_request(sockfd, buf, len);
+			parameter_values[2] = &inject_buffer;
+			parameter_values[3] = &inject_len;
 
-		ret = real_send(sockfd, buf, len, flags);
+			enable_inject = 1;
+		} else
+			rtr_http_sniff_request(sockfd, buf, len);
+	}
+
+	retrace_log_and_redirect_before(&event_info);
 
+	if (!err) {
+		ret = real_send(sockfd,
+			enable_inject ? inject_buffer : buf,
+			enable_inject ? inject_len : len,
+			flags);
 		if (errno)
```

Comment:
I've noticed a few places where `if (errno)` is used regardless of the result of the previous system call. My understanding is that the value of `errno` is undefined unless the previous call indicates a failure. In this case `if (ret == -1)` would be a better choice as we're not actually bothered about the value of `errno`